shaders/src/light_reflections.fs

```diff
@@ -0,0 +1,269 @@
+
+    // We only advance the z field of Q in the inner loop, since Q.xy is never used until after the
+    // loop terminates.
+
+    for (Point2 P = P0;
+        ((P.x * stepDirection) <= end) &&
+        (stepCount < maxSteps) &&
+        ((rayZMax < sceneZMax - vsZThickness) ||
+            (rayZMin > sceneZMax)) &&
+        (sceneZMax != 0.0);
+        P += dP, Q.z += dQ.z, k += dk, stepCount += 1.0) {
+
+        hitPixel = permute ? P.yx : P;
+
+        // The depth range that the ray covers within this loop iteration.  Assume that the ray is
+        // moving in increasing z and swap if backwards.  Because one end of the interval is shared
+        // between adjacent iterations, we track the previous value and then swap as needed to
+        // ensure correct ordering
+        rayZMin = prevZMaxEstimate;
+
+        // Compute the value at 1/2 pixel into the future
+        rayZMax = (dQ.z * 0.5 + Q.z) / (dk * 0.5 + k);
+        prevZMaxEstimate = rayZMax;
+        if (rayZMin > rayZMax) { swap(rayZMin, rayZMax); }
+
+        // View-space z of the background
+        sceneZMax = linearizeDepth(texelFetch(vsZBuffer, int2(hitPixel), 0).r);
+    } // pixel on ray
+
+    Q.xy += dQ.xy * stepCount;
+    vsHitPoint = Q * (1.0 / k);
+    // Matches the new loop condition:
+    return (rayZMax >= sceneZMax - vsZThickness) && (rayZMin <= sceneZMax);
+}
+
+#undef Point3
+#undef Point2
+#undef Vector2
+#undef Vector3
+#undef Vector4
+
+// -- end "BSD 2-clause license" -------------------------------------------------------------------
+
+mat4x4 scaleMatrix(float x, float y) {
+    mat4x4 m = mat4(1.0f);
+    m[0].x = x;
+    m[1].y = y;
+    m[2].z = 1.0f;
+    m[3].w = 1.0f;
+    return m;
+}
+
+mat4x4 translateMatrix(float x, float y) {
+    mat4x4 m = mat4(1.0f);
+    m[3].xy = vec2(x, y);
+    return m;
+}
+
+// random number between 0 and 1, using interleaved gradient noise
+float ssr_random(const highp vec2 w) {
+    const vec3 m = vec3(0.06711056, 0.00583715, 52.9829189);
+    return fract(m.z * fract(dot(w, m.xy)));
+}
+
+void evaluateScreenSpaceReflections(highp vec3 r, inout vec3 Fr) {
+    highp vec3 wsRayDirection = r;
+    highp vec3 wsRayStart = shading_position + frameUniforms.ssrBias * wsRayDirection;
+
+    // ray start/end in view space
+    highp vec4 vsRayStart = getViewFromWorldMatrix() * vec4(wsRayStart, 1.0);
+    highp vec4 vsRayDirection = getViewFromWorldMatrix() * vec4(wsRayDirection, 0.0);
+
+    highp vec3 vsOrigin = vsRayStart.xyz;
+    highp vec3 vsDirection = vsRayDirection.xyz;
+    highp float vsZThickness = frameUniforms.ssrThickness;
+    // TODO: use the actual near plane.
+    highp float nearPlaneZ = -0.1f;
+    highp float stride = 1.0;
+    // TODO: jitterFraction should be a between 0 and 1, but anything < 1 gives banding artifacts.
+    highp float jitterFraction = 1.0f;
+    // TODO: set this to the larger of the viewport dimensions.
+    highp float maxSteps = 1000.0;
```

Comment:
This seems way too high. Should it be user parameterized as well?
And do all those values need to be `highp`?

Reply:
@romainguy no, I think this is just the actual maximum to can possibly do before it exists the viewport. It's not the number of step that will actually be used (I think).

Reply:
Right, this is just a fail-safe to avoid tracing further than necessary. The equivalent user param is `ssrDistance`. And yes, some of these values don't need to be `highp`.